modules/030-cloud-provider-yandex/hooks/yandex_preemptibly_delete_preemtible_instances.go

```diff
@@ -0,0 +1,244 @@
+	if err != nil {
+		return nil, fmt.Errorf("can't access class name of Machine %q: %s", obj.GetName(), err)
+	}
+	if len(className) == 0 {
+		return nil, fmt.Errorf("spec.class.name is empty in %q", obj.GetName())
+	}
+
+	return &Machine{
+		Name:              obj.GetName(),
+		CreationTimestamp: obj.GetCreationTimestamp(),
+		Terminating:       terminating,
+		MachineClassKind:  classKind,
+		MachineClassName:  className,
+	}, nil
+}
+
+func isPreemptibleFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	preemptible, ok, err := unstructured.NestedBool(obj.UnstructuredContent(), "spec", "preemptible")
+	if err != nil {
+		return nil, fmt.Errorf("can't access field \"preemptible\" of YandexInstanceClass %q: %s", obj.GetName(), err)
+	}
+
+	preemptible = ok && preemptible
+
+	return &YandexInstanceClass{
+		Name:        obj.GetName(),
+		Preemptible: preemptible,
+	}, nil
+}
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
+	Queue:        "/modules/cloud-provider-yandex/preemtibly-delete-preemtible-instances",
+	Schedule: []go_hook.ScheduleConfig{
+		{
+			Name:    "every-15",
+			Crontab: "0/15 * * * *",
+		},
+	},
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:                   "machines",
+			ExecuteHookOnEvents:    go_hook.Bool(false),
+			WaitForSynchronization: go_hook.Bool(false),
+			ApiVersion:             "machine.sapcloud.io/v1alpha1",
+			Kind:                   "Machine",
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"d8-cloud-instance-manager"},
+				},
+			},
+			FilterFunc: applyMachineFilter,
+		},
+		{
+			Name:                   "ics",
+			ExecuteHookOnEvents:    go_hook.Bool(false),
+			WaitForSynchronization: go_hook.Bool(false),
+			ApiVersion:             "deckhouse.io/v1",
+			Kind:                   "YandexInstanceClass",
+			FilterFunc:             isPreemptibleFilter,
+		},
+	},
+}, deleteMachines)
+
+func deleteMachines(input *go_hook.HookInput) error {
+	timeNow := time.Now().UTC()
+
+	var (
+		icsSnapshot                   = input.Snapshots["ics"]
+		machineSnapshot               = input.Snapshots["machines"]
+		preemptibleInstanceClassesSet = make(map[string]struct{})
+		machines                      []*Machine
+	)
+
+	for _, icRaw := range icsSnapshot {
+		ic, ok := icRaw.(*YandexInstanceClass)
+		if !ok {
+			return fmt.Errorf("failed to assert to *YandexInstanceClass")
+		}
+
+		if ic.Preemptible {
+			preemptibleInstanceClassesSet[ic.Name] = struct{}{}
+		}
+	}
+
+	for _, machineRaw := range machineSnapshot {
+		machine, ok := machineRaw.(*Machine)
+		if !ok {
+			return fmt.Errorf("failed to assert to *Machine")
+		}
+
+		if machine.Terminating {
+			continue
+		}
+
+		if machine.MachineClassKind != "YandexMachineClass" {
+			continue
+		}
+
+		splittedMachineClassName := strings.Split(machine.MachineClassName, "-")
```

Comment:
It looks like a magic conversion that can be accidentally broken in the future. Adding a comment is required.